src/vm/util.cpp

```diff
@@ -3281,144 +3281,231 @@ INT32 InternalCasingHelper::InvariantToLowerNoThrow(__out_bcount_opt(cMaxBytes)
+
 // Convert szIn to lower case in the Invariant locale.
 INT32 InternalCasingHelper::InvariantToLowerHelper(__out_bcount_opt(cMaxBytes) LPUTF8 szOut, int cMaxBytes, __in_z LPCUTF8 szIn, BOOL fAllowThrow)
 {
-
-    CONTRACTL {
+    CONTRACTL
+    {
         // This fcn can trigger a lazy load of the TextInfo class.
         if (fAllowThrow) THROWS; else NOTHROW;
-        if (fAllowThrow) GC_TRIGGERS; else GC_NOTRIGGER;
+        GC_TRIGGERS;
         if (fAllowThrow) {INJECT_FAULT(COMPlusThrowOM());} else {INJECT_FAULT(return 0);}
         MODE_ANY;
 
         PRECONDITION((cMaxBytes == 0) || CheckPointer(szOut));
         PRECONDITION(CheckPointer(szIn));
-    } CONTRACTL_END
+    }
+    CONTRACTL_END
 
-    int inLength = (int)(strlen(szIn)+1);
     INT32 result = 0;
 
-    LPCUTF8 szInSave = szIn;
-    LPUTF8 szOutSave = szOut;
-    BOOL bFoundHighChars=FALSE;
-    //Compute our end point.
-    LPCUTF8 szEnd;
-    INT32 wideCopyLen;
+    LPUTF8 szStart = szOut;
+    LPUTF8 szEnd = szOut + cMaxBytes - 1;
 
-    CQuickBytes qbOut;
-    LPWSTR szWideOut;
+    INT32 c;
 
-    if (cMaxBytes != 0 && szOut == NULL) {
-        if (fAllowThrow) {
+    if (cMaxBytes != 0 && szOut == NULL)
+    {
+        if (fAllowThrow)
+        {
             COMPlusThrowHR(ERROR_INVALID_PARAMETER);
         }
         SetLastError(ERROR_INVALID_PARAMETER);
-        result = 0;
         goto Exit;
     }
 
     if (cMaxBytes) {
-        szEnd = szOut + min(inLength, cMaxBytes);
-        //Walk the string copying the characters.  Change the case on
-        //any character between A-Z.
-        for (; szOut<szEnd; szOut++, szIn++) {
-            if (*szIn>='A' && *szIn<='Z') {
-                *szOut = *szIn | 0x20;
-            }
-            else {
-                if (((UINT32)(*szIn))>((UINT32)0x80)) {
-                    bFoundHighChars = TRUE;
-                    break;
+        //Walk the string copying the characters.
+        while (szOut<szEnd && *szIn != 0) {
+            if (((UINT32)(*szIn)) > ((UINT32)0x80))
+            {
+                //Multi-byte code point
+                szIn = Utf8DecodeMultibyteChar(szIn, &c);
+                if (!szIn)
+                {
+                    RaiseException(ERROR_NO_UNICODE_TRANSLATION, EXCEPTION_NONCONTINUABLE, 0, 0);
                 }
-                *szOut = *szIn;
-            }
-        }
 
-        if (!bFoundHighChars) {
-            //If we copied everything, tell them how many bytes we copied,
-            //and arrange it so that the original position of the string + the returned
-            //length gives us the position of the null (useful if we're appending).
-            if (--inLength > cMaxBytes) {
-                if (fAllowThrow) {
-                    COMPlusThrowHR(HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER));
+                // Convert WCHAR character to lower case
+                // Skip doing this for the supplementary plane characters because we would do the wrong thing anyway
```

Comment:
>Skip doing this for the supplementary plane characters because we would do the wrong thing anyway [](start = 19, length = 97)

what is the wrong thing we'll do anyway when encountering Surrogates? would the behavior here will be different than the reflection?